### PR TITLE
feat(runtime, deployer)!: extract load phase + provisioning to deployer, resource update

### PR DESCRIPTION
**Pull request by jonaro00**
## Description of change

Resource update:
- Replaced `ResourceBuilder` with `ResourceConfigBuilder`, a new and updated API for writing plugins.

Breaking:
- Local run and deploy with outdated runtime will fail.
- DeploymentMetadata: removed field service_name. Use field project_name.
- These changes have been made compatible with console, only old c-s will break when parsing newer models
  - DatabaseResource is now serde untagged
  - SecretStore is serde transparent
- Old secrets will be invalidated due to being serde untagged (next deploy after upgrade will need fresh secrets file)

### CODE REVIEW

File: cargo-shuttle/src/lib.rs [hunk 1]
```diff
@@ -1103,17 +1090,101 @@ impl Shuttle {
...
+                                    .into_inner()
+                                    .into(),
+                                ),
+                            };
+                            shuttle_resources.push(resource::Response {
+                                r#type: shuttle_resource.r#type,
+                                config: serde_json::Value::Null,
+                                data: serde_json::to_value(&res).unwrap(),
+                            });
+                            *r = serde_json::to_vec(&ShuttleResourceOutput {
```
Comment (chesedo) [inline cargo-shuttle/src/lib.rs:1138]: Just comment why we are doing this

File: cargo-shuttle/src/lib.rs [hunk 2]
```diff
@@ -1103,17 +1090,101 @@ impl Shuttle {
...
         }
 
-        let resources = response
-            .resources
-            .into_iter()
-            .map(resource::Response::from_bytes)
-            .collect();
+        //
+        // PROVISIONING PHASE
+        //
```
Comment (chesedo) [inline cargo-shuttle/src/lib.rs:1095]: Please refactor all this into a separate function :smile:

File: cargo-shuttle/src/lib.rs [hunk 3]
```diff
@@ -1103,17 +1090,101 @@ impl Shuttle {
...
+                            shuttle_resources.push(resource::Response {
+                                r#type: shuttle_resource.r#type,
+                                config: serde_json::Value::Null,
+                                data: serde_json::Value::Null,
+                            });
+                        }
+                        resource::Type::Container => {
+                            let config = serde_json::from_value(shuttle_resource.config)
+                                .context("deserializing resource config")?;
+                            let res = prov.start_container(config).await?;
```
Comment (chesedo) [inline cargo-shuttle/src/lib.rs:1166]: This loop currently does a number of things:
1. It adds resources to `shuttle_resources`
2. It sometimes changes the resource
3. It sometimes provisions a resource

It is currently hard to reason about all of these. Can you somehow use something like a functional style to better separate these three tasks

File: codegen/src/shuttle_main/mod.rs [hunk 4]
```diff
@@ -226,74 +226,78 @@ impl ToTokens for Loader {
...
         };
 
-        let loader = quote! {
-            async fn loader(
-                mut #factory_ident: ::shuttle_runtime::__internals::ProvisionerFactory,
-                mut #resource_tracker_ident: ::shuttle_runtime::__internals::ResourceTracker,
-            ) -> #return_type {
+        let loader_runner = quote! {
+            async fn __loader(
+                #factory_ident: ::shuttle_runtime::__internals::ProvisionerFactory,
```
Comment (chesedo) [inline codegen/src/shuttle_main/mod.rs:256]: Should this still have provisioner in the name?

File: deployer/src/deployment/run.rs [hunk 5]
```diff
@@ -269,43 +277,191 @@ impl Built {
...
+        match serde_json::from_slice::<ResourceInput>(r.as_slice())
+            .context("deserializing resource input")?
+        {
+            ResourceInput::Shuttle(shuttle_resource) => {
+                macro_rules! log {
+                    ($msg:expr) => {
+                        info!("[Resource][{}] {}", shuttle_resource.r#type, $msg);
+                    };
+                }
+                macro_rules! get_cached_output_or_provision {
```
Comment (chesedo) [inline deployer/src/deployment/run.rs:340]: It's weird to have the macro rules here. Just move it to outside the function.

File: deployer/src/deployment/run.rs [hunk 6]
```diff
@@ -269,43 +277,191 @@ impl Built {
...
+                                let cached_output = resource.data.clone();
+                                log!("Found cached output");
+                                match serde_json::from_value::<$output_type>(cached_output) {
+                                    Ok(output) => Some(output),
+                                    Err(_) => {
+                                        log!("Failed to validate cached output");
+                                        None
+                                    }
+                                }
+                            });
```
Comment (chesedo) [inline deployer/src/deployment/run.rs:358]: I don't think this needs to be a macro. You should be able to refactor it into a function instead.

File: deployer/src/deployment/run.rs [hunk 7]
```diff
@@ -269,43 +277,191 @@ impl Built {
...
+                        resources_to_save.push(record_request::Resource {
+                            r#type: shuttle_resource.r#type.to_string(),
+                            config: serde_json::to_vec(&serde_json::Value::Null).unwrap(),
+                            data: serde_json::to_vec(&new_secrets).unwrap(),
+                        });
+                        *r = serde_json::to_vec(&ShuttleResourceOutput {
+                            output: new_secrets.clone(),
+                            custom: shuttle_resource.custom,
+                        })
+                        .unwrap();
```
Comment (chesedo) [inline deployer/src/deployment/run.rs:426]: Same comment as earlier about each loop iteration doing too much. It should be possible to simplify this.

File: deployer/src/lib.rs [hunk 8]
```diff
@@ -39,6 +39,9 @@ pub async fn start(
         .deployment_updater(persistence.clone())
         .resource_manager(persistence.clone())
         .builder_client(builder_client)
+        .provisioner_client(Some(
```
Comment (chesedo) [inline deployer/src/lib.rs:42]: This should not be an `Option`

File: proto/runtime.proto [hunk 9]
```diff
@@ -16,17 +16,16 @@ service Runtime {
...
-  // Name of service to load
-  string service_name = 1;
+  string project_name = 1;
+  // "local" or "deployment"
+  string env = 3;
+  // Secrets that belong to this deployment
+  map<string, string> secrets = 20;
 
-  // Path to compiled file to load for service
+  // deprecated
```
Comment (chesedo) [inline proto/runtime.proto:25]: There should be a better way to mark this :smile:

File: resources/qdrant/src/lib.rs [hunk 10]
```diff
@@ -41,62 +34,89 @@ impl Qdrant {
...
             },
         }
     }
 }
 
+#[derive(Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum Wrapper {
+    One(ShuttleResourceOutput<ContainerResponse>),
+    Two(QdrantClientConfigWrap),
```
Comment (chesedo) [inline resources/qdrant/src/lib.rs:95]: Just name these better :sweat_smile:

File: runtime/Cargo.toml [hunk 11]
```diff
@@ -15,7 +15,7 @@ doctest = false
 [dependencies]
 shuttle-codegen = { workspace = true, features = ["frameworks"] }
 shuttle-common = { workspace = true, features = ["backend", "claims"] }
-shuttle-proto = { workspace = true, features = ["provisioner", "runtime"] }
+shuttle-proto = { workspace = true, features = ["runtime"] }
```
Comment (chesedo) [inline runtime/Cargo.toml:18]: :partying_face:

File: runtime/src/alpha/mod.rs [hunk 12]
```diff
@@ -4,48 +4,44 @@ use std::{
...
 
-use self::args::Args;
+use crate::args::args;
 
-mod args;
+// uses custom macro instead of clap to reduce dependency weight
+args! {
+    pub struct Args {
+        // The port to open the gRPC control layer on.
+        // The address to expose for the service is given in the StartRequest.
```
Comment (chesedo) [inline runtime/src/alpha/mod.rs:36]: This comment is no longer correct right?

File: service/src/lib.rs [hunk 13]
```diff
@@ -57,37 +42,27 @@ pub trait Factory: Send + Sync {
...
-
-    /// The input config to this resource.
-    type Config: Serialize;
-
-    /// The output from requesting this resource.
-    /// A cached copy of this will be used if the same [`ResourceBuilder::Config`] is found for this [`ResourceBuilder::TYPE`].
-    type Output: Serialize + DeserializeOwned;
-
-    /// Get the config of this plugin after it has been built from its macro arguments with the builder pattern.
+pub trait IntoResourceInput: Default {
```
Comment (chesedo) [inline service/src/lib.rs:45]: We should keep this as `ResourceBuilder` and double down on naming the method `build()`

**Review by chesedo — COMMENTED** [review]
This is taking great shape. Thanks J!!

File: deployer/src/deployment/run.rs (hunk 5 above)
Comment (jonaro00) [reply deployer/src/deployment/run.rs:340]: For some reason it just did not work when I placed them there ...

File: runtime/src/alpha/mod.rs (hunk 12 above)
Comment (jonaro00) [reply runtime/src/alpha/mod.rs:36]: it is, the deployer just has the start input hardcoded to `0.0.0.0:8000`

File: service/src/lib.rs (hunk 13 above)
Comment (jonaro00) [reply service/src/lib.rs:45]: It does not build a resource, it (builds and) emits a request(+config) for a resource.
`IntoResourceRequest` is not good since it can also do a non-request (input := request | custom struct)

File: deployer/src/deployment/run.rs (hunk 6 above)
Comment (jonaro00) [reply deployer/src/deployment/run.rs:358]: my unfinished attempt: https://github.com/shuttle-hq/shuttle/pull/1628/commits/8c726055943354f5e070f790fe46e7b020eff436
feel free to try it out

File: service/src/lib.rs (hunk 13 above)
Comment (chesedo) [reply service/src/lib.rs:38]: `ResourceConfigBuilder` then :smile:

File: deployer/src/deployment/run.rs (hunk 5 above)
Comment (jonaro00) [reply deployer/src/deployment/run.rs:382]: Neither works :skull:

File: service/src/lib.rs (hunk 13 above)
Comment (jonaro00) [reply service/src/lib.rs:38]: How bout `ResourceInputBuilder` and `build() -> Result<Self::Input...`? :pleading_face:

File: deployer/src/deployment/run.rs [hunk 14]
```diff
@@ -269,94 +277,242 @@ impl Built {
...
-    debug!(shuttle.project.name = %service_name, shuttle.service.name = %service_name, "service loaded");
+async fn load(
+    service_name: String,
+    mut runtime_client: runtime::Client,
+    new_secrets: &HashMap<String, String>,
+) -> Result<Vec<Vec<u8>>> {
+    debug!(shuttle.project.name = %service_name, "loading service");
+    let response = runtime_client
+        .load(Request::new(LoadRequest {
+            project_name: service_name.clone(),
```
Comment (iulianbarbu) [inline deployer/src/deployment/run.rs:508]: Why does the load request need this?

File: resources/qdrant/src/lib.rs [hunk 15]
```diff
@@ -41,62 +34,89 @@ impl Qdrant {
...
+                    )
+                )),
             },
         }
     }
 }
 
+#[derive(Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum Wrapper {
```
Comment (iulianbarbu) [inline resources/qdrant/src/lib.rs:93]: nit: would be better to name it `OutputWrapper`?

File: resources/shared-db/src/postgres.rs [hunk 16]
```diff
@@ -19,51 +20,27 @@ impl Postgres {
...
+            serde_json::to_value(&self.0).unwrap(),
+            serde_json::Value::Null,
+        ))
     }
 }
 
 #[derive(Serialize, Deserialize)]
 #[serde(transparent)]
-pub struct Wrapper(DatabaseResource);
+pub struct Wrapper(ShuttleResourceOutput<DatabaseResource>);
```
Comment (iulianbarbu) [inline resources/shared-db/src/postgres.rs:38]: nit: same, `OutputWrapper`.

File: common/src/resource.rs [hunk 17]
```diff
@@ -38,7 +82,8 @@ pub enum Type {
     Database(database::Type),
     Secrets,
     Persist,
-    Custom,
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply common/src/resource.rs:41]: Not relevant anymore I guess. Resolving.

File: deployer/src/deployment/run.rs [hunk 18]
```diff
@@ -269,94 +277,242 @@ impl Built {
...
-                match serde_json::from_value::<SecretStore>(r.data.clone()) {
-                    Ok(ss) => {
-                        // Combine old and new, but insert old first so that new ones override.
-                        let mut combined = HashMap::from_iter(ss.into_iter());
-                        combined.extend(secrets.clone().into_iter());
-                        secrets = combined;
+                if shuttle_resource.version != RESOURCE_SCHEMA_VERSION {
+                    bail!("
+                        Shuttle resource request for {} with incompatible version found. Expected {}, found {}. \
+                        Make sure that this deployer and the Shuttle resource are up to date.
```
[the remark this answers is not in the document]
Comment (jonaro00) [reply deployer/src/deployment/run.rs:414]: When we change any part of the json schema

File: deployer/src/deployment/run.rs [hunk 19]
```diff
@@ -269,94 +277,242 @@ impl Built {
...
+                }
+
+                // TODO?: Make the version integer be part of the cached output
+
+                match shuttle_resource.r#type {
+                    resource::Type::Database(db_type) => {
+                        // no config fields are used yet, but verify the format anyways
+                        let _config: DbInput =
+                            serde_json::from_value(shuttle_resource.config.clone())
+                                .context("deserializing resource config")?;
```
[the remark this answers is not in the document]
Comment (jonaro00) [reply deployer/src/deployment/run.rs:429]: I'm signaling how the variable would be used

File: resources/qdrant/src/lib.rs (hunk 15 above)
Comment (jonaro00) [reply resources/qdrant/src/lib.rs:93]: It's essentially a private type that no one will look at. Does not really matter.

File: cargo-shuttle/tests/integration/run.rs [hunk 20]
```diff
@@ -54,20 +54,7 @@ async fn axum_static_files() {
...
-        .await
-        .unwrap()
-        .text()
-        .await
-        .unwrap();
-
-    assert!(
-        request_text.contains("This is an example of serving static files with axum and shuttle.")
-    );
+    assert!(request_text.contains("This is an example of serving"));
```
Comment (chesedo) [inline cargo-shuttle/tests/integration/run.rs:57]: Why did this test change?

File: deployer/src/deployment/run.rs [hunk 21]
```diff
@@ -272,100 +281,251 @@ impl Built {
             .await
             .map_err(Error::Runtime)?;
 
-        kill_old_deployments.await?;
-        // Execute loaded service
-        load(
+        kill_old_deployments.await?; // TODO: Move to between load and provision? Or even after provision?
```
Comment (chesedo) [inline deployer/src/deployment/run.rs:284]: Todo

File: runtime/src/alpha/mod.rs [hunk 22]
```diff
@@ -108,16 +101,12 @@ pub async fn start(loader: impl Loader<ProvisionerFactory> + Send + 'static) {
     // where to serve the gRPC control layer
     let addr = SocketAddr::new(Ipv4Addr::LOCALHOST.into(), args.port);
 
-    let provisioner_address = args.provisioner_address;
     let mut server_builder = Server::builder()
         .http2_keepalive_interval(Some(Duration::from_secs(60)))
-        .layer(JwtAuthenticationLayer::new(AuthPublicKey::new(
```
Comment (chesedo) [inline runtime/src/alpha/mod.rs:114]: Can we also remove the `claims` feature on `shuttle-common`?

File: service/src/lib.rs [hunk 23]
```diff
@@ -20,35 +21,12 @@ pub mod error;
 #[cfg(feature = "runner")]
 pub mod runner;
 
-/// An interface for the provisioner used in [`ResourceBuilder::output`].
-#[async_trait]
-pub trait Factory: Send + Sync {
```
Comment (chesedo) [inline service/src/lib.rs:25]: Finally!! :partying_face:

File: cargo-shuttle/src/lib.rs [hunk 24]
```diff
@@ -1150,6 +1153,106 @@ impl Shuttle {
...
+                    if shuttle_resource.version != RESOURCE_SCHEMA_VERSION {
+                        bail!("
+                            Shuttle resource request for {} with incompatible version found. Expected {}, found {}. \
+                            Make sure that cargo-shuttle and the Shuttle resource are up to date.
+                            ",
+                            shuttle_resource.r#type,
+                            RESOURCE_SCHEMA_VERSION,
+                            shuttle_resource.version
+                        );
+                    }
```
Comment (chesedo) [inline cargo-shuttle/src/lib.rs:1188]: Can we repeat the same improvements we made yesterday here?

File: cargo-shuttle/tests/integration/run.rs (hunk 20 above)
Comment (jonaro00) [reply cargo-shuttle/tests/integration/run.rs:57]: The served HTML page changed in examples

File: deployer/src/deployment/run.rs (hunk 21 above)
Comment (jonaro00) [reply deployer/src/deployment/run.rs:284]: Do you think it should happen after the load phase or after the provision phase?

File: deployer/src/deployment/run.rs [hunk 25]
```diff
@@ -272,100 +281,251 @@ impl Built {
...
+                        shuttle_resource.r#type,
+                        RESOURCE_SCHEMA_VERSION,
+                        shuttle_resource.version
+                    ))
                 }
+            }).collect::<anyhow::Result<Vec<_>>>()?.into_iter()
+    {
+        // Mutate resource bytes with provisioning output if relevant
+
+        // TODO?: Make the version integer be part of the cached output
```
[the remark this answers is not in the document]
Comment (jonaro00) [reply deployer/src/deployment/run.rs:428]: Did we agree that this should remain a TODO until it is moved into provisioner?

File: deployer/src/deployment/run.rs (hunk 21 above)
Comment (chesedo) [reply deployer/src/deployment/run.rs:284]: After provisioning... or as late as possible.

File: deployer/src/deployment/run.rs (hunk 25 above)
[the remark this answers is not in the document]
Comment (chesedo) [reply deployer/src/deployment/run.rs:428]: We can do that. Then it should not be a question mark `todo?` though

**Review by chesedo — APPROVED** [review]
LGTM

File: .circleci/config.yml [hunk 26]
```diff
@@ -790,14 +790,14 @@ workflows:
     jobs:
       - approve-deploy-images-unstable:
           type: approval
-          filters:
-            branches:
-              only: main
+          # filters:
+          #   branches:
```
Comment (iulianbarbu) [inline .circleci/config.yml:794]: nit: a reminder to remove these before merging.

File: examples
Comment (iulianbarbu) [inline examples:1]: Are these changes required in this PR?

[the remark this answers is not in the document]
Comment (jonaro00) [reply examples:1]: It's just a bump to a more recent commit on main. One of the tests depends on it

**Review by iulianbarbu — APPROVED** [review]

**Review by oddgrd — COMMENTED** [review]
I didn't get through all of it yet, but don't delay merging on my account, I am quite late to the party :smile: I left a few nits/minor comments, but LGTM so far!

File: cargo-shuttle/src/provisioner_server.rs [hunk 27]
```diff
@@ -19,19 +19,15 @@ use futures::StreamExt;
...
+    Ping, Pong,
 };
 use shuttle_service::database::Type;
-use tokio::{task::JoinHandle, time::sleep};
-use tonic::{
-    transport::{self, Server},
-    Request, Response, Status,
-};
+use tokio::time::sleep;
+use tonic::{Request, Response, Status};
```
[the remark this answers is not in the document]
Comment (jonaro00) [reply cargo-shuttle/src/provisioner_server.rs:30]: It is needed for the client
